### PR TITLE
[FEATURE] Post Update Command

### PR DESCRIPTION
**Pull request by developStorm**
Support to execute a shell command after a successful update, e.g. to restart the server by RCON.

### CODE REVIEW

**Comment by Owen-Cochell** [comment]
Hello again!

I am not sure about this feature. This poses some development questions, which I will be listing below. But my main problem does not lie in implementing these changes, but in the necessity of this feature. The point of this script(At least in my mind) is to update the server in an efficient, automated way, nothing more, nothing less. I don't think that it is necessary for this script to have a feature that runs an external program, as that falls outside of the purpose of this script. If one is using this script in an environment that promotes automation, then whatever system they are using to execute this script can easily be configured to run whatever arbitrary post-update command that they see fit. I am also worried that this will greatly increase the code base in size and complexity, as I want this application to be a standalone, all in one script that is small and self contained. I don't want it becoming too big or unwieldy, as all of the source code is in a singular file. Plus, from a security standpoint, this feature can be potentially dangerous. We do not know what a user's environment looks like, or how it is configured, and having this feature COULD(Massive could here), be a security risk. But most of all, I do not this it is necessary, relevant feature that we need to add.

Now, I am always open for debate, and I am interested to hear your spin on this. In the meantime, Here are some development questions/requests I would like implemented:

1. Use the subprocess module instead of os.system(). The subprocess module is a more robust, secure method of executing system calls, as stated here[https://docs.python.org/3/library/os.html#os.system] in the documentation, and was actually created to replace the older os.system() functions.

2. Your argpase configuration does not collect extra arguments, as it only takes one value. For example, if I executed the following:
>python server_update.py --post-command shutdown_server.sh --timeout 70 [PATH]

Argparse would only associate 'shutdown_server.sh' with the --post-command argument(I will make a comment on this in your code as well, so you can see where this problem lies). To alleviate this, we must specify the number of arguments argpase is expecting, and how it will store those arguments. This can be done using the nargs parameter, and can be read about here[https://docs.python.org/3/library/argparse.html#nargs]. We would also have to split this content into a list, which is the format that subprocess takes. 

3. Another problem is overlapping parameters. For example, this following command would trigger an argaprse failure:
>python server_update.py -v 15.3.2 --post-command server_shutdown.sh -v [PATH]

The -v is associated with the server_shutdown.sh command, but argparse would see it as a command associated with us, as -v is an argument we can expect to see, meaning that at the very best argparse will replace our intended value(15.3.2) with a potentially irrelevant value, and at the very worst, a argparse failure. Putting a limit on what arguments a external command can have seems very restrictive, and at that point, users will likely use their own methods of executing the script outside of our own, making this implementation pointless.

4. How would we handle input/output/errors? If the program we want to run outputs text to stdout, should we output that to the terminal? To a file? Somewhere else? Let the user decide? It's a tricky situation. If we let the user decide the output location/type, then that will add a good amount of code to our code base, as we would have to handle multiple output scenarios. On the flip side, if we simply choose an output format/type, then that may not be robust enough for users, meaning that they won't be using this feature anyway. That's just for output, what about input? Errors that occur? How would we log/handle those? Again we would have to add a lot of code to make this feature robust and worth using, and we also made the command line usage more complicated, as we would have to add a lot of new arguments the user would have to specify.

These are just some development questions I would like answered, as well as my argument against adding this feature. Now, I am completely open for debate, and would love to hear what you think about all this! If you have a counter argument, found an error in my analysts, or have found something that I may have missed, I would be happy to hear it. In the meantime, this is my stance on the issue.

File: server_update.py [hunk 1]
```diff
@@ -1029,6 +1029,7 @@ def get_new(self, default_version='latest', default_build='latest'):
     parser.add_argument('-i', '--interactive', help='Prompts the user for the version they would like to install.', action='store_true')
     parser.add_argument('-nlc', '--no-load-config', help='Will not load config information', action='store_false')
     parser.add_argument('-ndc', '--no-dump-config', help="Will not dump configuration information.", action='store_false')
+    parser.add_argument('-p', '--post-update-command', help="Will execute this command once the update is completed.", default="NONE")
```
Comment (Owen-Cochell) [inline server_update.py:1035]: This argument only takes one parameter, when it should take multiple. This can be done using the nargs parameter.

**Comment by developStorm** [comment]
I didn't use `argparse ` before so I honestly don't know `nargs` 😓 . However, I think using `nargs` here might not benefits much since we can simply use a pair of quotes to make the command as one parameter. For example:

``` bash
python server_update.py -v 15.3.2 --post-command "server_shutdown.sh -v [PATH]"
```

The double quote and single quote both work in this situation so we don't need to worry if the user needs to use quotes within the command (since they can always choose to use another kind of quote or even use the escape sequence).

This also makes the Issue 3 invalid -- we restricted the range of the external command between quotes.

For issue 4, Input/output/errors with `os.system` and `subprocess.call` go to the same pipe as where these outputs of `server_update.py` go. I think this is enough since we don't actually have an option for users to specify the output location, therefore it seems too early to consider has this option for external commands. Users can also decide this by themselves by doing like 

``` bash
python server_update.py -v 15.3.2 --post-command "server_shutdown.sh -v [PATH]" > /var/log/server_update.log 2>&1
```

For issue 1, I'll go ahead and replace that with `subprocess.call`. A list of args is not necessary since in the document you give, they are using string concentration -- `+` to combine command and args, which is meanless when we get the whole command as one string.

**Comment by Owen-Cochell** [comment]
Hey again!

Valid points, was not aware argparse was able to group by quotes. That solves our command line problem. However, using your solution to redirect output to a file also prevents the user from seeing output from the script, which isn't a problem for some, but defiantly conflicts with the '--interactive' flag.  Again, their are some fancy command line solutions for redirection, but at that point you might as well have the program manage output itself. I also agree with your sentiment that adding support for subprocess output to an external location would also need support for normal application output.

Again, I just think that this feature is outside of the scope for this project. If you are working in an automated environment, it would be easier and more robust to call a external program using your own methods. Same goes for a interactive shell. I just don't think that we can give users a reason to use this feature without making an impact on the codebase.

Again, this is completely my opinion. Perhaps you have some insight that I do not on the matter, and if you do, I would love to hear it! I am totally open for any solutions to some of the above problems, as well as a counter argument to my claims.

**Pull request closed without merging**